Add Build-CI 2.0

dougiesquire opened the pull request:
This PR adds Build-CI. Because there is already quite extensive CI testing inherited from the upstream repo, I have only added manifests to build access-om3 using oneapi with MOM6 `+/~asymmetric_mem`.

Closes #20

dougiesquire committed:
Add Build-CI 2.0

anton-seaice commented:
I'm sure you know this but you'll need to set 

`find_package(MPI REQUIRED COMPONENTS Fortran)` in CMakeLists and make the required / related changes:

(e.g similar to  https://github.com/ACCESS-NRI/access3-share/commit/d7d7aacaff18737d98717b428bdcae1b59555b1e )

(actually maybe that's only for the standalone build)

CodeGat commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,39 @@
+  pull_request:
+  push:
+#     branches:
+#       - [0-9]*.[0-9]* # Default branch naming
```
The branches syntax isn't quite regex - it's globs. So it'd be `- '*.*'` instead of the regex you have - see https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax#onpull_requestpull_request_targetbranchesbranches-ignore

dougiesquire replied:
Yeah. Weirdly this syntax works for branch protections...

dougiesquire replied:
[These docs](https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax#filter-pattern-cheat-sheet) suggest what I had should work?

dougiesquire replied:
Just needed to be quoted: `'[0-9]*.[0-9]*'`

dougiesquire committed:
Don't reuse specs from buildcache for packages with infinity default versions

dougiesquire committed:
Try Spack v1.0

dougiesquire committed:
Try using spack-config dougie-delete branch

dougiesquire committed:
Try using spack-packages dougie-delete branch

dougiesquire committed:
Use CMake@3.31.6

dougiesquire committed:
Empty commit to test deps reuse

dougiesquire commented:
Hmm, I pushed an empty commit to test that dependencies are being reused from the build cache, but it appears to have rebuilt everything (?). Ping @CodeGat in case you're interested (not expecting you to do anything here)

dougiesquire committed:
Another empty commit to test new caches

dougiesquire committed:
Use spack-config:main and spack-packages:api-v2 now that fixes are implemented there

dougiesquire committed:
Another empty commit after hotfix to spack-config:main

dougiesquire commented:
Latest failure is due to missing language dependencies in the access3-share SPR. Fix is here: https://github.com/ACCESS-NRI/spack-packages/pull/303

dougiesquire committed:
Another empty commit after fix to spack-packages:api-v2

CodeGat committed:
Tommy's empty commit to check if the logging is fixed

dougiesquire committed:
Use spack-packages dougie-delete branch

The branch replaces the "stable" version of access3_share with:
version("develop", branch="develop", preferred=True)
so that I can push changes to this branch and test whether the package rebuilds

dougiesquire committed:
Empty commit now that I've actually made an access3_share develop branch

dougiesquire committed:
Empty commit to confirm reuse of access3_share@develop when there are no changes to branch

dougiesquire commented:
The build-ci on the previous commit _did_ reuse `access3_share` from the cache - see [here](https://github.com/ACCESS-NRI/MOM6/actions/runs/16963340801/job/48081157566#step:13:3713)

I'll now make a change on the `access3_share` `develop` branch and then check whether or not the cache is still used.

dougiesquire committed:
Now with modified access3_share develop branch

I've also excluded the symmetric build manifest for clarity

dougiesquire commented:
Hmmm, it appears to have used the cache :'( - see [here](https://github.com/ACCESS-NRI/MOM6/actions/runs/16963894978/job/48082947660#step:13:3713)

It's correctly showing the updated commit in the [concretized output](https://github.com/ACCESS-NRI/MOM6/actions/runs/16963894978/job/48082947660#step:13:675)

dougiesquire commented:
In the concretized output the commit variant changes but the spack hash does not:

[Pre update](https://github.com/ACCESS-NRI/MOM6/actions/runs/16963340801/job/48081157566#step:13:675) to develop branch:
```
  -   hdbt3sf          ^access3-share@develop~ipo~openmp build_system=cmake build_type=Release commit=f2f35ce5915e82a83899b69560d826deab53b668 generator=make arch=linux-rocky8-x86_64_v3 %c,fortran=oneapi@2025.2.0
```

[Post update](https://github.com/ACCESS-NRI/MOM6/actions/runs/16963894978/job/48082947660#step:13:675) to develop branch:
```
  -   hdbt3sf          ^access3-share@develop~ipo~openmp build_system=cmake build_type=Release commit=33c5bd243a7ca2263fabd1330429816de4ff2426 generator=make arch=linux-rocky8-x86_64_v3 %c,fortran=oneapi@2025.2.0
```

dougiesquire commented:
Oh, it looks like one fix might be to add `no_cache=True` to our branch-based `version()` directives - see [here](https://spack.readthedocs.io/en/latest/packaging_guide_creation.html#download-caching) (also available in v0.22). I'll test this tomorrow, though it would still be better if we could have caching/reuse that recognised commit hash changes on branch-based versions.

dougiesquire committed:
Now with no_cache=True on access3_share develop version

dougiesquire commented:
Weird, even with `no_cache=True` it still appears to have used the cache - see [here](https://github.com/ACCESS-NRI/MOM6/actions/runs/16978072277/job/48131917260?pr=24#step:13:3727).

@CodeGat, I wonder if this is because there was already an entry in the cache before I set `no_cache=True`? Is it possible/easy to remove a specific package (`hdbt3sf`) from the cache? If not, I can try testing with a different package.

anton-seaice commented:
Did the spack hash change ?

spack directives (e.g. version()) tend to impact concretisation, rather than install. So I wonder if `no_cache` means don't use the cache for concretisation ? But if it concretises the same as something in the cache, it will still use the build-cache for a binary?

Now that i've written that, you probably added a commit between installs for testing anyway right ?

dougiesquire commented:
> Did the spack hash change ?
> 
> spack directives (e.g. version()) tend to impact concretisation, rather than install. So I wonder if `no_cache` means don't use the cache for concretisation ? But if it concretises the same as something in the cache, it will still use the build-cache for a binary?
> 
> Now that i've written that, you probably added a commit between installs for testing anyway right ?

See [this comment above](https://github.com/ACCESS-NRI/MOM6/pull/24#issuecomment-3188217804)

dougiesquire committed:
Now with hdbt3sf removed from the build cache

dougiesquire commented:
Ok great, that did _not_ use the cache - see [here](https://github.com/ACCESS-NRI/MOM6/actions/runs/16979429108/job/48136063612#step:13:3972)

I'll now make another change on the `access3_share` `develop` branch and then check whether or not the cache is still used.

dougiesquire committed:
Now with modified access3_share develop branch

dougiesquire commented:
Again, the concretization [shows](https://github.com/ACCESS-NRI/MOM6/actions/runs/16979771135/job/48137126654#step:13:675) the changed commit
```
 -   hdbt3sf          ^access3-share@develop~ipo~openmp build_system=cmake build_type=Release commit=326f5b611b748d3f8e911d0b9999e22dbc36db9d generator=make arch=linux-rocky8-x86_64_v3 %c,fortran=oneapi@2025.2.0
 ```
 
but the cache is still used at install - see [here](https://github.com/ACCESS-NRI/MOM6/actions/runs/16979771135/job/48137126654#step:13:3699)

aidanheerdegen commented:
> Because there is already quite extensive CI testing inherited from the upstream repo

Are those checks only run when you do PRs back to upstream?

dougiesquire commented:
> Are those checks only run when you do PRs back to upstream?

Nah they're run on `push`/`pull_request`. I've just turned them off for now in this PR to reduce noise

aidanheerdegen commented:
Did you mean to test with `mpich` rather than openmpi?

https://github.com/ACCESS-NRI/MOM6/actions/runs/16978072277/job/48131917260#step:13:722
```
 -   5tr3xy7          ^mpich@3.4.3~argobots~cuda+fortran+hwloc+hydra~level_zero+libxml2+pci~rocm+romio~slurm~verbs+wrapperrpath~xpmem build_system=autotools datatype-engine=auto device=ch4 netmod=ofi patches:=66ba8eb,7326028,be529bf,db4173c,de0de41 pmi=default arch=linux-rocky8-x86_64_v3 %c,cxx,fortran=oneapi@2025.2.0
```

dougiesquire committed:
Add ESMF spec to manifest

Testing this issue with Spack v1.0: https://github.com/ACCESS-NRI/ACCESS-OM3/issues/12#issuecomment-3195101604

dougiesquire commented:
> Did you mean to test with `mpich` rather than openmpi?

Nope.

dougiesquire committed:
Use openmpi

dougiesquire commented:
> spack.error.ConfigError: Version requirement 2025.03: on fms for fms cannot match any known version from package.py or externals

But this concretized with `fms@2025.03` two commits ago... Possibly the `spack` `releases/v1.0` branch is changing under our feet?

ADDED: Or more likely, the `spack-config` `main` branch.

ADDED ADDED: yup - changes from yesterday reference [this commit](https://github.com/ACCESS-NRI/spack-config/blob/67d1dd31e5e5023ce70c27175a2eaf3884549fbb/v1.0/all/repos.yaml#L7C13-L7C18), which is prior to the addition of the 2025 FMS versions

ADDED ADDED ADDED: fix [here](https://github.com/ACCESS-NRI/spack-config/pull/66), now merged

dougiesquire committed:
Remove FMS spec from manifest

dougiesquire committed:
Empty commit to trigger rebuild

dougiesquire committed:
And again

dougiesquire committed:
And again again

dougiesquire committed:
And again again again

dougiesquire commented:
@aidanheerdegen, @anton-seaice it's definitely not a conclusive test, but with the following now in the manifest

```
    esmf:
      require:
        - '@8.7.0'
        - 'fflags="-march=sapphirerapids -mtune=sapphirerapids -unroll -O0"'
        - 'cflags="-march=sapphirerapids -mtune=sapphirerapids -unroll -O0"'
        - 'cxxflags="-march=sapphirerapids -mtune=sapphirerapids -unroll -O0"'
```

Spack v1.0 has concretized ESMF with the [same flag order and spack hash](https://github.com/ACCESS-NRI/MOM6/actions/runs/17056358164/job/48354705416#step:13:683) 5 times and reused ESMF from the cache:
```
 -   vc4gct5          ^esmf@8.7.0 cflags='-march=sapphirerapids -mtune=sapphirerapids -unroll -O0' cxxflags='-march=sapphirerapids -mtune=sapphirerapids -unroll -O0' fflags='-march=sapphirerapids -mtune=sapphirerapids -unroll -O0' ~debug~external-lapack+external-parallelio+mpi+netcdf~pnetcdf~python+shared~xerces build_system=makefile esmf_comm=auto esmf_os=auto esmf_pio=auto patches:=f63d405 snapshot=none arch=linux-rocky8-x86_64_v3 %c,cxx,fortran=oneapi@2025.2.0
 ```

CodeGat commented:
> But this concretized with `fms@2025.03` two commits ago... Possibly the `spack` `releases/v1.0` branch is changing under our feet?

If you want to verify this (for future reference), you can check the last line of the `Update spack` step, which will give something like:
```
Repository was updated from f9cc209ab9955d635c78e4ae6e1d7866e6cf7cc3 to 17d34285729c1bb39dfbdef09c5b78d88bb9e3f9
```

It's also an output of the job (`outputs.spack-sha`)

CodeGat committed:
Test using the :rocky-v1.0 image

CodeGat committed:
TEST: Use `api-v2` branch

CodeGat committed:
Retest with v1.0 spack ref and dougie-delete branch again

dougiesquire committed:
Use spack-packages branch api-v2

dougiesquire commented:
@CodeGat we can stop using the `dougie-delete` branch of spack-packages - that was just for testing. But it still fails during the Spack setup to load a compiler:

```
Loading compiler: gcc@13.2.0 %gcc@8.5.0 target=x86_64
==> Error: Spec 'gcc@13.2.0 %gcc@8.5.0 arch=None-None-x86_64' matches no installed packages.
```

This didn't used to happen. Possibly @harshula has an idea what's going on?

CodeGat commented:
Nah that one is on me. I'm investigating it now.

CodeGat committed:
Test out :rocky-v1.0-2025.09.001

CodeGat committed:
Second test with filled up buildcache

CodeGat committed:
Final test after upstream spack rebuilt

CodeGat committed:
TEST (for fun): try loading compilers via upstream lockfile rather than manifest, and use bootstrapped image

CodeGat committed:
Actual final test after bootstrap merged into build-ci

CodeGat commented:
@dougiesquire sorry about all the noise, I'm satisfied that it works now :laughing: - also made it about 2 minutes faster by bootstrapping spack in the image rather than doing it in the running container. 

With regard to https://github.com/ACCESS-NRI/MOM6/pull/24/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR33 - would you be able to remove that and push (so I can then review the PR as a whole), if that's what you want to do for this PR?

dougiesquire committed:
Add back removed manifest; use openmpi

dougiesquire commented:
Apologies @CodeGat, I had to change the base branch and rebase.

dougiesquire committed:
Add back upstream CI workflows

dougiesquire commented:
Hmmm, I tried to force it to use `openmpi` (previously was using `mpich`) and [got](https://github.com/ACCESS-NRI/MOM6/actions/runs/17752008326/job/50448424080#step:14:121):

```
==> [2025-09-16-02:01:48.543202] Error: failed to concretize `access-om3@latest ^openmpi` for the following reasons:
     1. cannot satisfy a requirement for package 'perl'.
```

CodeGat commented:
Just rerun it and yep, that's an error :laughing: I'm not really sure why, though. I've got it down to a minimal example, at least...
```
[root@77064e1cd29a /]# spack spec openmpi
==> Error: failed to concretize `openmpi` for the following reasons:
     1. cannot satisfy a requirement for package 'perl'.
[root@77064e1cd29a /]# spack spec openmpi%intel@2021.10.0
==> Error: failed to concretize `openmpi %intel@2021.10.0` for the following reasons:
     1. cannot satisfy a requirement for package 'perl'.
[root@77064e1cd29a /]# spack spec openmpi%oneapi@2025.2.0
==> Error: failed to concretize `openmpi %oneapi@2025.2.0` for the following reasons:
     1. cannot satisfy a requirement for package 'perl'.
[root@77064e1cd29a /]# spack spec openmpi%gcc
==> Error: failed to concretize `openmpi %gcc` for the following reasons:
     1. cannot satisfy a requirement for package 'perl'.
```

I don't really know what this means though. Something to do with `spack-config`?

The image is using ACCESS-NRI/spack-packages@83407dee31885f7c3919eb35d4a7e991151829cc and ACCESS-NRI/spack-config@e7592522e4ec66f7547fd686f87a2ed95aa1b6f6

dougiesquire commented:
I don't know why it can't satisfy the requirement. @harshula may have insights?

CodeGat commented:
We've found the source of the issue. `perl` is classed as an external non-buildable package and `spack 1.0` can't ascertain what compiler was used to build it. Which means that when we enforce compiler constraints (`%oneapi`) it doesn't know if that install of `perl` satisfies that constraint. The way to fix it is to make `perl` a buildable package so it knows what compiler was used. I'll open a PR in `spack-config` to fix this, and test that branch in this PR.

CodeGat commented:
@dougiesquire, depending on how long these verification checks take, you might have to change the parallelism on them...they currently reserve all 20 of the organisations GitHub-hosted runner jobs to do them...

dougiesquire commented:
Yeah, those are just inherited from upstream. They shouldn't take more than a few mins each but also happy to limit how many can run at once.

CodeGat commented:
Making `perl` buildable seems to have gotten it to concretize, which is good! If it works, I'll get the `spack-config` PR merged and revert the `spack-config-ref` to `main`

harshula commented:
Hi, When using Spack v1, the compiler constraint is set per language. e.g. https://github.com/ACCESS-NRI/dev-docs/wiki/Spack#spack-v10-caveats , as per https://github.com/ACCESS-NRI/spack-config/issues/74#issuecomment-3302084156

CodeGat committed:
Use spack-config branch 74-perl-buildable to sidestep upstream externals unknown compiler

CodeGat commented:
@dougiesquire, sorry about all the noise. @harshula and myself have been testing out using the spack-v1.0-specific language requirements over just making `perl` a non-external package (as that is a more robust solution). 

Unfortunately, that solution doesn't work when we have a `1.0` runner reading from a `0.22` upstream in `build-ci`, so we're going to have to go with the `perl` solution until we have moved allllllllllll the components to `spack 1.0` as well, so we can move the upstream to `1.0` too.

I think once the checks complete, this will be good to merge.

dougiesquire commented:
No worries @CodeGat - thanks for resolving this!

We're in our usual review stalemate. But if you could approve to formally indicate you're happy to merge then I'll bypass the rules.
